apps/rush-lib/src/cli/scriptActions/BulkScriptAction.ts

```diff
@@ -182,36 +207,66 @@ export class BulkScriptAction extends BaseScriptAction {
           ' operating system and number of CPU cores.'
       });
     }
-    this._toFlag = this.defineStringListParameter({
+    this._toProject = this.defineStringListParameter({
       parameterLongName: '--to',
       parameterShortName: '-t',
+      argumentName: 'PROJECT2',
+      description:
+        'Restricts command execution to the specified project and all dependencies of the specified project. ' +
+        '"." can be used as shorthand to specify the project in the current working directory. ' +
+        'Specifying "--to" or "--to-except" multiple times will take the union of matching projects. ' +
+        'Specifying "--from" as well will further restrict execution to the intersection of both filters.',
```

Comment:
> 'Specifying "--from" as well will further restrict execution to the intersection of both filters.',

Suppose we have this graph:

```
A --[depends-on]--> B  --[depends-on]--> C
```

I would expect this command to build all projects:
```shell
$ rush build --to B --to-except B --from B --from-except B
```

However the description above seems to imply that the set will be "restricted" to some sort of intersection.  (?)

Would appending additional CLI parameters /ever/ reduce the set of projects to be built? If so that seems counterintuitive.

Reply:
@octogonz , with the modification, the command:
```
$ rush build --to B --from B
```
will build `B` and only `B` regardless of the shape of the graph.

Similarly,
```
$ rush build --to B --from A
```
matches 0 projects because:
`--to B` matches `B, C`
`--from A` matches `A`
And the intersection of these two is the empty set.

The issue this is addressing is that currently if you have the following graph:
```
A --[depends-on]--> B  --[depends-on]--> C --[depends-on]--> D
```
And you want to rebuild only `B` and `C`, there is no way to ask for that to happen.
With the change, you can do this with:
```
$ rush build --from C --to B
```
i.e. build everything from `C` to `B`

I suppose my question would be, right now, does anyone combine `--from` and `--to` in build commands? I never do because the combination is completely useless.

Reply:
> I suppose my question would be, right now, does anyone combine `--from` and `--to` in build commands? I never do because the combination is completely useless.

I use this all the time.  For example, suppose that I made a change to API Extractor:
```
# 1. Build all the dependencies needed by API Documenter
# 2. Build API Documenter itself
# 3. Build only projects that depend on API Documenter to make sure no tests were broken
$ rush build --to api-documenter --from api-documenter
```

This select 19 projects, which is much faster than building the full 106 projects in this repo.

The behavior you described is a significant design change, that would cause the above command to do something not useful.

Reply:
I'm also curious how your design would handle more complex graphs.  For example, consider:
```shell
$ rush build --from A --to H
```
![1](https://user-images.githubusercontent.com/4673363/105115579-eab3dd00-5a7d-11eb-9412-39d843fb5aaa.png)

With the current design, Rush would build the union of everything needed to build H, plus everything that depends on A.

It sounds like your design would instead compute an intersection like this:

![2](https://user-images.githubusercontent.com/4673363/105115672-16cf5e00-5a7e-11eb-99e3-c96a7badc6cb.png)

Is that meaningful?  What does it represent exactly?

Reply:
FYI PR #2349 considered it a bug that Rush does not build B in this example:

```shell
$ rush build --from A
```

![3](https://user-images.githubusercontent.com/4673363/105115858-788fc800-5a7e-11eb-8510-bef043a32051.png)

E may fail to compile because B as skipped.  So PR #2349 fixes this problem by ensuring that B will get included.

Perhaps there are two competing mentalities here:

1. *"Give me a quick way to skip compiling irrelevant projects, but do not expect me to remember how projects are chained together or whether they need to be rebuilt"* VERSUS
2. *"I save time by optimistically assuming most projects are up to date; if I'm wrong then I'll get a weird build error, but no problem, I can just do a full rebuild"*

Reply:
Here's another interesting case:
```shell
# Build only these 3 projects and their dependencies
$ rush build --to D --to E --to F
```
Will this still work with your new design?

Reply:
I'm revising the PR to separate out adding support for `--from-except` and `--to-except` and will revisit adding a way to express "rebuild the minimal set of projects required to reflect changes to A in B" as a follow-on. As long as the machinery exists under the hood, the watch mode machinery doesn't really care if there's a way to ask for that set from the command line.

Reply:
Closing for now since the change to behavior of `--from` and `--to` has been reverted. Will revisit in a later issue, since it clearly needs more discussion.